ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+			return Ok(state);
+		}
+		// BlockState is not found in memory, which means we need to reconstruct state from last checkpoint.
+		match self.client.read().as_ref().and_then(|w| { w.upgrade() }) {
+			None => {
+				return Err(From::from("failed to upgrade client reference"));
+			}
+			Some(c) => {
+				let last_checkpoint_number = (header.number() / self.epoch_length as u64) * self.epoch_length;
+				assert_ne!(last_checkpoint_number, header.number());
+
+				let mut chain: &mut VecDeque<Header> = &mut VecDeque::with_capacity(
+					(header.number() - last_checkpoint_number + 1) as usize);
+
+				// Put ourselves in.
+				chain.push_front(header.clone());
+
+				// populate chain to last checkpoint
+				let mut last = chain.front().ok_or(
+					"just pushed to front, reference must exist;")?.clone();
+
+				while last.number() != last_checkpoint_number + 1 {
+					match c.block_header(BlockId::Hash(*last.parent_hash())) {
+						None => {
+							return Err(Box::new(
+								format!("parent block {} of {} could not be recovered.",
+										&last.parent_hash(), &last.hash())
+							).into());
+						}
+						Some(next) => {
+							chain.push_front(next.decode()?.clone());
+							last = chain.front().ok_or(
+								"just pushed to front, reference must exist;")?.clone();
+						}
+					}
+				}
+
+				// Catching up state, note that we don't really store block state for intermediary blocks,
+				// for speed.
+				let backfill_start = time::Instant::now();
+				info!(target: "engine",
+						"Back-filling block state. last_checkpoint_number: {}, target: {}({}).",
+						last_checkpoint_number, header.number(), header.hash());
+
+				// Get the state for last checkpoint.
+				let last_checkpoint_hash = *(chain.front().ok_or(
+					"just pushed to front, reference must exist; qed"
+				)?.parent_hash());
+				let last_checkpoint_header = match c.block_header(BlockId::Hash(last_checkpoint_hash)) {
+					None => return Err(From::from("Unable to find last checkpoint block")),
+					Some(header) => header.decode()?,
+				};
+
+				let last_checkpoint_state: CliqueBlockState;
+
+				// We probably don't have it cached, but try anyway.
+				if let Some(st) = block_state_by_hash.get_mut(&last_checkpoint_hash) {
+					last_checkpoint_state = (*st).clone();
+				} else {
+					last_checkpoint_state = self.new_checkpoint_state(&last_checkpoint_header)?;
+				}
+				block_state_by_hash.insert(last_checkpoint_header.hash().clone(), last_checkpoint_state.clone());
+
+				// Backfill!
+				let mut new_state = last_checkpoint_state.clone();
+				for item in chain {
+					new_state.apply(item, false)?;
+				}
+				new_state.calc_next_timestamp(header, self.period);
+				block_state_by_hash.insert(header.hash(), new_state.clone());
+
+				let elapsed = backfill_start.elapsed();
+				info!(target: "engine",
+						"Back-filling succeed, took {} ms.",
+						elapsed.as_secs() * 1000 + elapsed.subsec_millis() as u64,
```

Comment:
This might overflow you have to cast it to `u128` to be on the safe-side.

Also can you add a comment that we should replace this with [Duration::as_millis()](https://doc.rust-lang.org/std/time/struct.Duration.html#method.as_millis) which will stabilized in `Rust 1.33`?